src/HDF5.jl

```diff
@@ -1370,12 +1381,65 @@ Base.getindex(parent::Union{File,Group}, r::Reference) = _deref(parent, r)
 Base.getindex(parent::Dataset, r::Reference) = _deref(parent, r) # defined separately to resolve ambiguity
 
 # convert special types to native julia types
-normalize_types(x) = x
-normalize_types(x::NamedTuple{T}) where {T} = NamedTuple{T}(map(normalize_types, values(x)))
-normalize_types(x::Cstring) = unsafe_string(x)
-normalize_types(x::FixedString) = unpad(String(collect(x.data)), pad(x))
-normalize_types(x::FixedArray) = normalize_types.(reshape(collect(x.data), size(x)...))
-normalize_types(x::VariableArray) = normalize_types.(copy(unsafe_wrap(Array, convert(Ptr{eltype(x)}, x.p), x.len, own=false)))
+function normalize_types(::Type{T}, buf::AbstractMatrix{UInt8}) where {T}
+    # First dimension spans bytes of a single element of type T --- (recursively) normalize
+    # each range of bytes to final type, returning vector of normalized data.
+    return [_normalize_types(T, view(buf, :, ind)) for ind in axes(buf, 2)]
+end
+
+# high-level description which should always work --- here, the buffer contains the bytes
+# for exactly 1 element of an object of type T, so reinterpret the `UInt8` vector as a
+# length-1 array of type `T` and extract the (only) element.
+function _typed_load(::Type{T}, buf::AbstractVector{UInt8}) where {T}
+    return @inbounds reinterpret(T, buf)[1]
+end
+# fast-path for common concrete types with simple layout (which should be nearly all cases)
+function _typed_load(::Type{T}, buf::V) where {T, V <: Union{Vector{UInt8}, Base.FastContiguousSubArray{UInt8,1}}}
+    dest = Ref{T}()
+    GC.@preserve buf Base._memcpy!(unsafe_convert(Ptr{Cvoid}, dest), pointer(buf), sizeof(T))
+    return dest[]
```

Comment:
```suggestion
    return GC.@preserve buf unsafe_load(convert(Ptr{T}, pointer(buf)))
```

@musm, does this optimize better one Windows? I see the same exact LLVM IR on Linux.

This is probably also simpler to read — I was motivated to use the `_memcpy!` version after digging through `ReinterpretArray` internals, but for this case where we're only loading a single element, `unsafe_load` is probably the better choice anyway.

Reply:
Actually, on second thought, there might be a semantic difference — x86 is forgiving about unaligned loads (so loading 8-byte `Int` from an address where address mod 8 != 0), but that's not true on all architectures. IIUC, `memcpy` is defined to work correctly no matter the alignment, but doing so might take several instructions if loads must be aligned.

I _think_ this shows that: https://godbolt.org/z/snvrMM

Reply:
This does optimize cleanly:

![image](https://user-images.githubusercontent.com/4319522/101378300-eee6bb00-3880-11eb-9bcf-59efd9247030.png)

where `__untyped_load` is the modified one above.

Reply:
Do you find something like
```ll
call void @llvm.memcpy.p0i8.p0i8.i64(i8* align 1 %20, i8* %17, i64 8, i1 false)
```
in the unoptimized LLVM IR? I'm assuming yes, so Julia is giving LLVM the info it needs, but then LLVM isn't optimizing it away.

Reply:
Yes that's precisely what I find. This isn't anything blocking, but I was curious why it doesn't optimize.

Reply:
@jmert I ran the test suite on x86 on Windows with the modified method:
```julia
function _typed_load(::Type{T}, buf::V) where {T, V <: Union{Vector{UInt8}, Base.FastContiguousSubArray{UInt8,1}}}
    return GC.@preserve buf unsafe_load(convert(Ptr{T}, pointer(buf)))
end
```
and the suite passed.

I can't comment on which is technically correct.

Reply:
My understanding was that Julia protects us from this:

> Int from an address where address mod 8 != 0)

https://docs.julialang.org/en/v1/devdocs/object/

at the end of the document

> Note that all objects are allocated in multiples of 4 bytes and aligned to the platform pointer size. Memory is allocated from a pool for smaller objects, or directly with malloc() for large objects.

Reply:
Once we've recursed into something like `_normalize_types(::Type{<:NamedTuple})` where we're taking arbitrary `view`s, Julia's allocation guarantee no longer applies to the alignment of any particular pointer we take.

Reply:
I think as long as the order is correct it shouldn't be an issue since Julia's member alignment is the same as C semantics.

Reply:
OK — how about using the `unsafe_load` version, with the `_memcpy!` variant commented out and waiting just in case it has to be resurrected?

Reply:
I'm on board with that if you are?  If we get bug reports on the next release, we can quickly release a fix, but the test suit is passing for me.

> x86 is forgiving about unaligned loads (so loading 8-byte Int from an address where address mod 8 != 0), but that's not true on all architectures. IIUC, memcpy is defined to work correctly no matter the alignment, but doing so might take several instructions if loads must be aligned.

> I think this shows that: https://godbolt.org/z/snvrMM

I'm not 100% confident that this won't be an issue, but I'm inclinced to think it isn't.

Reply:
If we are very concerend, It might be worth opening a discourse question to confirm these assumptions and see what others say.